File: web/src/state/sync.ts
/**
 * Keeping the Continue shelf the same on every machine.
 *
 * The local database stays the source of truth for the machine it is on.
 * This is an addition to it, never a replacement — everything here can fail,
 * and a player whose sync fails is a player that works exactly as it did
 * before any of this existed. That is the safety property the whole design
 * hangs on, and it is why nothing below is allowed to throw.
 *
 * The channel is behind an interface rather than reached for directly. The
 * decisions worth getting right — when to send, whether anything changed,
 * what to do with what comes back — are here and are tested against a fake.
 * Only the MTProto calls are left untested, and they are a thin adapter in
 * `telegram/state-channel.ts`.
 */

import { mergeStates } from "./merge";
import { parseRecord, type SyncRecord } from "./sync-record";
import type { WatchState } from "./store";

/** One device's document, as it sits on the channel. */
export interface ChannelDocument {
  messageId: number;
  /** Whose it is, from the caption. */
  device: string;
  /** The JSON body. */
  text: string;
}

/**
 * Where the documents live.
 *
 * Deliberately small: list them, and put mine. A device only ever writes its
 * own message — Telegram has no compare-and-swap, so two devices editing one
 * shared message would clobber each other with no way to notice.
 */
export interface StateChannel {
  list(): Promise<ChannelDocument[]>;
  /** Sends, or edits `messageId` when this device has written before. */
  put(body: string, messageId: number | null): Promise<number>;
}

export interface SyncOutcome {
  /** Rows this machine took in. */
  pulled: number;
  /** Whether a document was actually sent. */
  pushed: boolean;
  /** What went wrong, if anything. Never thrown. */
  failed?: string;
}

export class StateSync {
  /** This device's own message, once it is known. */
  private mine: number | null = null;
  /** The last body sent, so an unchanged one is not sent again. */
  private lastSent: string | null = null;
  /** The shared drain in progress. Never rejects: `round` catches. */
  private running: Promise<SyncOutcome> | null = null;
  private again = false;

  constructor(
    private readonly state: WatchState,
    private readonly channel: StateChannel,
    private readonly device: string,
  ) {}

  /**
   * Read everyone's, merge, apply selected changes, and write back. Calls
   * arriving during a round share its drain and retain one follow-up read.
   *
   * Pull before push so what this machine sends already reflects what it just
   * learnt. A device that pushed first would publish a document it knew to be
   * out of date, and every other device would have to do the reconciling that
   * this one had the information to do.
   */
  once(): Promise<SyncOutcome> {
    // A burst needs one follow-up, not one queued read per event. News that
    // arrives during that follow-up asks for another; no active read can
    // consume an event announcing something it has already missed.
    if (this.running) {
      this.again = true;
      return this.running;
    }
    this.running = (async () => {
      const outcome: SyncOutcome = { pulled: 0, pushed: false };
      try {
        do {
          this.again = false;
          const round = await this.round();
          outcome.pulled += round.pulled;
          outcome.pushed ||= round.pushed;
          if (round.failed !== undefined) outcome.failed = round.failed;
        } while (this.again);
        return outcome;
      } finally {
        this.running = null;
      }
    })();
    // Every caller, including shutdown, waits for all retained news. The
    // result reports work and failures from the whole shared drain.
    return this.running;
  }

  private async round(): Promise<SyncOutcome> {
    let pulled = 0;
    try {
      const documents = await this.channel.list();
      pulled = this.mergeChannelDocuments(documents);
      const pushed = await this.pushIfChanged();
      return { pulled, pushed };
    } catch (error) {
      // Swallowed on purpose. A channel that cannot be reached costs a log
      // line, and the player carries on against its own database.
      return { pulled, pushed: false, failed: describe(error) };
    }
  }

  /**
   * Merges what the channel holds into this machine.
   *
   * This device contributes its current local export. Its channel document
   * only recovers the message ID to edit on the next push; merging that stale
   * copy back in could restore state the viewer has since removed locally.
   */
  private mergeChannelDocuments(documents: ChannelDocument[]): number {
    const records: SyncRecord[] = [this.state.exportRecord(this.device)];
    for (const document of documents) {
      // A device's own message is recognised here rather than filtered out, so
      // that the id is learnt even on a run where nothing needs sending.
      if (document.device === this.device) this.mine = document.messageId;
      else {
        const record = parseRecord(document.text);
        if (record !== null) records.push(record);
      }
    }
    return this.state.importMerged(mergeStates(records));
  }

  /** Writes this machine's document, unless it would be the same one again. */
  private async pushIfChanged(): Promise<boolean> {
    const record = this.state.exportRecord(this.device);
    // `writtenAt` changes on every export and nothing reads it during a merge,
    // so it is left out of the comparison: including it would make every
    // document different from the last and upload one on every tick for ever.
    const body = JSON.stringify(record);
    const comparable = JSON.stringify({ ...record, writtenAt: 0 });
    if (comparable === this.lastSent) return false;

    this.mine = await this.channel.put(body, this.mine);
    this.lastSent = comparable;
    return true;
  }
}

function describe(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: crates/mediagram-core/src/state/sync.rs
//! Keeping the Continue shelf the same on every device.
//!
//! Matches `web/src/state/sync.ts`: local state remains authoritative, and
//! sync failures never prevent playback. [`StateChannel`] separates the
//! merge and retry decisions from the Telegram adapter so tests can exercise
//! complete rounds without a live account.

use std::collections::HashMap;

mod device;
mod error;
pub use device::device_id;
use error::SyncError;

use super::StateDb;
use super::exchange;
use super::merge::merge_states;
use super::record::{SyncRecord, parse_record};

/// One device's document, as it sits on the channel.
#[derive(Debug, Clone, PartialEq)]
pub struct ChannelDocument {
    pub message_id: i32,
    /// Whose it is, from the caption.
    pub device: String,
    /// The JSON body.
    pub text: String,
}

/// Where the documents live.
///
/// Deliberately small: list them, and put mine. A device only ever writes
/// its own message — Telegram has no compare-and-swap, so two devices
/// editing one shared message would clobber each other with no way to
/// notice.
pub trait StateChannel {
    type Error: std::fmt::Display;
    async fn list(&self) -> Result<Vec<ChannelDocument>, Self::Error>;
    /// Sends, or edits `message_id` when this device has written before.
    async fn put(&self, body: String, message_id: Option<i32>) -> Result<i32, Self::Error>;
}

#[derive(Debug, Clone, Default, PartialEq, uniffi::Record)]
pub struct SyncOutcome {
    /// Rows this device took in.
    pub pulled: u64,
    /// Whether a document was actually sent.
    pub pushed: bool,
    /// What went wrong, if anything. Never a panic, never a `Result`.
    pub failed: Option<String>,
}

/// What a round remembers between calls, keyed by library handle so that
/// switching the chosen library never reuses another channel's message id
/// or "nothing changed" memory — see `api::state_sync` on why a handle
/// switch must not touch either.
#[derive(Default)]
pub struct SyncMemo {
    by_handle: HashMap<String, Memo>,
}

impl SyncMemo {
    pub fn entry(&mut self, handle: &str) -> &mut Memo {
        self.by_handle.entry(handle.to_string()).or_default()
    }
}

/// This device's own message on one channel, once it is known, and the
/// last body sent there so an unchanged one is not sent again.
#[derive(Default, Clone)]
pub struct Memo {
    mine: Option<i32>,
    last_sent: Option<String>,
}

/// Runs a complete round under the same lock used by the production adapter.
/// A timer and a channel event must not both send this device's first document.
pub async fn serialized<C: StateChannel>(
    memo: &tokio::sync::Mutex<SyncMemo>,
    handle: &str,
    state_db: &StateDb,
    channel: &C,
    device: &str,
) -> SyncOutcome {
    let mut memo = memo.lock().await;
    once(state_db, channel, device, memo.entry(handle)).await
}

/// One round: read everyone's, merge, take in what is newer, write back.
///
/// Pull before push so what this device sends already reflects what it
/// just learnt. A failed import rolls back; a failed send keeps the imported
/// rows and reports their count so callers can refresh the local library.
pub async fn once<C: StateChannel>(
    state_db: &StateDb,
    channel: &C,
    device: &str,
    memo: &mut Memo,
) -> SyncOutcome {
    let mut outcome = SyncOutcome::default();
    if let Err(failed) = round(state_db, channel, device, memo, &mut outcome).await {
        outcome.failed = Some(failed.to_string());
    }
    outcome
}

async fn round<C: StateChannel>(
    state_db: &StateDb,
    channel: &C,
    device: &str,
    memo: &mut Memo,
    outcome: &mut SyncOutcome,
) -> Result<(), SyncError<C::Error>> {
    let documents = channel.list().await.map_err(SyncError::Channel)?;
    outcome.pulled =
        merge_and_import_documents(state_db, &documents, device, memo).ok_or(SyncError::Import)?;
    outcome.pushed = publish_state_if_changed(state_db, channel, device, memo).await?;
    Ok(())
}

/// Merges what the channel holds into this device.
///
/// **This device's own document is part of the merge** — built fresh from
/// `state.db` rather than trusted from what the channel handed back, so
/// nothing here depends on this device having ever sent one. Without it
/// the merge would answer only what the others knew, and `import_merged`
/// is corrective rather than wholesale precisely so that this cannot erase
/// anything — but including it is what makes the answer complete rather
/// than merely safe.
fn merge_and_import_documents(
    state_db: &StateDb,
    documents: &[ChannelDocument],
    device: &str,
    memo: &mut Memo,
) -> Option<u64> {
    state_db.with(|conn| {
        let mut records: Vec<SyncRecord> = vec![exchange::export_record(conn, device)?];
        for document in documents {
            // A device's own message is recognised here rather than
            // filtered out beforehand, so its id is learnt even on a
            // round where nothing needs sending.
            if document.device == device {
                memo.mine = Some(document.message_id);
            } else if let Some(record) = parse_record(&document.text) {
                records.push(record);
            }
        }
        exchange::import_merged(conn, &merge_states(&records))
    })
}

/// Writes this device's document, unless it would be the same one again.
async fn publish_state_if_changed<C: StateChannel>(
    state_db: &StateDb,
    channel: &C,
    device: &str,
    memo: &mut Memo,
) -> Result<bool, SyncError<C::Error>> {
    let record = state_db
        .with(|conn| exchange::export_record(conn, device))
        .ok_or(SyncError::Read)?;
    let body = serde_json::to_string(&record)?;
    // `writtenAt` changes on every export and nothing reads it during a
    // merge, so it is left out of the comparison: including it would make
    // every document different from the last and send one on every tick
    // for ever.
    let mut comparable_record = record.clone();
    comparable_record.written_at = 0.0;
    let comparable = serde_json::to_string(&comparable_record)?;
    if memo.last_sent.as_deref() == Some(comparable.as_str()) {
        return Ok(false);
    }

    let sent = channel
        .put(body, memo.mine)
        .await
        .map_err(SyncError::Channel)?;
    memo.mine = Some(sent);
    memo.last_sent = Some(comparable);
    Ok(true)
}

#[cfg(test)]
#[path = "sync_tests.rs"]
mod tests;
